Made attackable positions customizable

reaperrr opened the pull request:
This allows modders to define custom offsets relative to `Actor.CenterPosition` that will be targeted instead of `CenterPosition`, which is extremely useful for tall units like mechs.

~~Also adds custom HitShapes to all TS units (not buildings for now, no point in that until we have footprint/custom shape support).~~ split to #11169.

Closes #7717.

reaperrr committed:
Refactor Health to serve as supplier of AttackablePositions

Additionally adds AttackablePositionOffset, allowing to customize the
position where attackers shoot at.

reaperrr committed:
Adapt HitShapes to custom AttackablePositions

Phrohdoh commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -65,6 +65,9 @@ public class ArmamentInfo : UpgradableTraitInfo, IRulesetLoaded, Requires<Attack
 		public WeaponInfo WeaponInfo { get; private set; }
 		public WDist ModifiedRange { get; private set; }
 
+		[Desc("Should the targets' custom TargetingPointZOffset be ignored.")]
```
`target's` is proper here.

Phrohdoh commented on OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs:
```diff
@@ -51,8 +51,20 @@ public void RenderAfterWorld(WorldRenderer wr, Actor self)
 				healthInfo.Shape.DrawCombatOverlay(wr, wcr, self);
 
+				var targetZc = Color.Magenta;
+				var tx = new WVec(96, 0, 0);
```
What is this 96 value and why 96?
Can this logically be customizable?

reaperrr replied:
> What is this 96 value and why 96?

Offset relative to TargetingCenterPosition, used to define the start and end points of the lines for drawing the cross overlay.

> Can this logically be customizable?

Yes. But nearly all other debug overlays we have aren't customizable, either.

Phrohdoh replied:
Okay, these look to be width and height values.

>Yes. But nearly all other debug overlays we have aren't customizable, either.

I find that to be a poor reason not to allow someone to customize something, but ok for consistency.

reaperrr replied:
I don't mean that I'm generally against making it customizable, I just didn't bother while working on this PR.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Health.cs:
```diff
@@ -19,9 +19,13 @@ public class HealthInfo : ITraitInfo, UsesInit<HealthInit>
 		[Desc("Trigger interfaces such as AnnounceOnKill?")]
 		public readonly bool NotifyAppliedDamage = true;
 
+		[Desc("Distance above CenterPosition for targeting by weapons.")]
```
I don't think this makes much sense (doesn't "scale").
Consider a east-to-west facing ship.
Would this make more sense as a worldly vector offset? (consider turret offset)

reaperrr replied:
The value is passed to a WVec's Z value later, so technically it already is.

reaperrr replied:
I was considering to make this directly WVec, but felt it would complicate matters (and increase risk of bugs in damage calculation etc.) if X and Y were customizable as well.

Phrohdoh replied:
I tend to agree that it may, but it also allows this feature to be used with actors whose graphics aren't perfectly vertical.

Imagine a Titan or similar with a weak spot that is offset from the centerpoint's Z axis.

GraionDilach replied:
Doesn't worth the hassle. If directional armors get in, a combo of that would work better.

Phrohdoh replied:
Directional armor doesn't at all relate to the projectile's initial target position.
The "hassle" is literally changing this to a `WVec` then changing the referencing code to use this directly instead of creating a `WVec` from components.

But ok, this can be done later when someone requests it.

reaperrr replied:
On the other hand, `Capsule` and `Rectangle` shapes can already be positioned along the Y axis, so they're not necessarily centered around the `Actor.CenterPosition`. So admittedly, allowing to fully customize the offset and not just Z would be more thorough and consistent.

I guess I'll give it a try.

reaperrr commented:
Updated. Fixed description and offset is now a full WVec, allowing custom X/Y position as well (updated TS Titan as showcase).

Phrohdoh commented on OpenRA.Game/Actor.cs:
```diff
@@ -52,6 +52,18 @@ public sealed class Actor : IScriptBindable, IScriptNotifyBind, ILuaTableBinding
+				if (health != null)
+					offset = health.Offset;
+
+				return CenterPosition + offset;
```
`return CenterPosition +  health == null ? WVec.Zero : health.Offset;`

Phrohdoh commented on OpenRA.Mods.Common/Effects/Missile.cs:
```diff
@@ -757,8 +757,10 @@ public void Tick(World world)
-					+ new WVec(WDist.Zero, WDist.Zero, info.AirburstAltitude);
+			{
+				var airburstVec = new WVec(WDist.Zero, WDist.Zero, info.AirburstAltitude);
+				newTarPos = (!args.IgnoreTargetPointOffset ? args.GuidedTarget.TargetingCenterPosition : args.GuidedTarget.CenterPosition) + airburstVec;
```
You can simplify that ternary.
`args.IgnoreTargetPointOffset ? args.GuidedTarget.CenterPosition : args.GuidedTarget.TargetingCenterPosition`

Phrohdoh commented on OpenRA.Mods.Common/HitShapes/Capsule.cs:
```diff
@@ -27,6 +27,12 @@ public class CapsuleShape : IHitShape
+		[Desc("Defines the top offset relative to the actors' target point.")]
+		public readonly int VerticalTopOffset = 0;
+
+		[Desc("Defines the bottom offset relative to the actors' target point.")]
```
`actor's` for both descriptions.

Phrohdoh commented on OpenRA.Mods.Common/HitShapes/Circle.cs:
```diff
@@ -22,11 +22,21 @@ public class CircleShape : IHitShape
 		[FieldLoader.Require]
 		public readonly WDist Radius = new WDist(426);
 
+		[Desc("Defines the top offset relative to the actors' target point.")]
```
`actor's` for both descriptions.

Phrohdoh commented on OpenRA.Mods.Common/HitShapes/Rectangle.cs:
```diff
@@ -25,10 +25,17 @@ public class RectangleShape : IHitShape
 		[FieldLoader.Require]
 		public readonly int2 BottomRight;
 
+		[Desc("Defines the top offset relative to the actors' target point.")]
```
`actor's` again

Phrohdoh commented on OpenRA.Mods.Common/HitShapes/Circle.cs:
```diff
@@ -35,12 +45,24 @@ public WDist DistanceFromEdge(WVec v)
+			var actorTargPos = actor.TargetingCenterPosition;
+
+			if (pos.Z <= actorTargPos.Z + VerticalTopOffset && pos.Z >= actorTargPos.Z + VerticalBottomOffset)
+				return DistanceFromEdge(pos - new WPos(actorTargPos.X, actorTargPos.Y, pos.Z));
```
Newlines please.

reaperrr replied:
Done.

Phrohdoh commented on OpenRA.Mods.Common/HitShapes/Rectangle.cs:
```diff
@@ -68,13 +86,24 @@ public WDist DistanceFromEdge(WVec v)
+			var actorTargPos = actor.TargetingCenterPosition;
+
+			if (pos.Z <= actorTargPos.Z + VerticalTopOffset && pos.Z >= actorTargPos.Z + VerticalBottomOffset)
+				return DistanceFromEdge((pos - new WPos(actorTargPos.X, actorTargPos.Y, pos.Z)).Rotate(-actor.Orientation));
```
Newlines, please.

Phrohdoh commented on OpenRA.Mods.Common/HitShapes/Rectangle.cs:
```diff
@@ -68,13 +86,24 @@ public WDist DistanceFromEdge(WVec v)
 		{
-			var verts = combatOverlayVerts.Select(v => wr.ScreenPosition(actor.CenterPosition + v.Rotate(actor.Orientation)));
-			wcr.DrawPolygon(verts.ToArray(), 1, Color.Yellow);
+			var vertsTo = combatOverlayVertsTop.Select(v => wr.ScreenPosition(actor.TargetingCenterPosition + v.Rotate(actor.Orientation)));
```
One more letter makes this a bit easier to understand at a glance.
Might as well change the below to `vertsBelow`, also.

reaperrr commented:
Updated.

pchote commented:
I'm really not a fan of having separate `CenterPosition` and `TargetCenterPosition`s.  CenterPosition being at the bottom of the actor is a bug, and should be fixed properly for all cases.

reaperrr commented:
> I'm really not a fan of having separate CenterPosition and TargetCenterPositions. CenterPosition being at the bottom of the actor is a bug, and should be fixed properly for all cases.

I see your point, but wouldn't that cause all kinds of regressions with pathfinding, attack distance and so on (at least in TS)? Problems like #5841 could become more common and I'm not sure I'd manage to find and fix them all (not on my own, in any case).

reaperrr commented:
I made a quick test locally, and among other things, `AnimationWithOffset` is positioned relative to `Actor.CenterPosition`. Generally speaking, In my opinion using `Actor.CenterPosition` for everything is just asking for trouble and regressions.

But more importantly, after some thinking I'd like to point out this:
If someone (maybe myself) where to add SW:Empire at War / Universe at War-style hit zones, then splitting targeting points from CenterPosition would become mandatory anyway.

My idea in that regard is to refactor the `Health` trait into a `HitPoint` of which an actor can have multiple ones, and only when health/HP of all (or at least all required) HitPoints has been reduced to 0 is the actor considered 'killed'.
I'd allow modders to assign `Armor` types to a specific `HitPoint`, give `HitPoint` a customizable `Priority` property (HitPoints with higher priority would be attacked first, if multiple HitPoints have the same priority, the attacker switches between them randomly), and a `RequiredForKill` property together with adding a property to warheads to ignore certain `HitPoint`s.
That could be used to implement StarCraft-style shields as well, among other nice things. 
Could come in quite handy for my MechCommander RTS mod as well. I guess I'll make a ticket.

pchote commented:
> But more importantly, after some thinking I'd like to point out this:
If someone (maybe myself) where to add SW:Empire at War / Universe at War-style hit zones, then splitting targeting points from CenterPosition would become mandatory anyway.

We already have a system in place for this: the `ITargetablePositions` interface and `Target.Positions`.

reaperrr commented:
Just wanted to note that I'm still working on this, using `ITargetablePositions` as suggested. It's just not ready yet.

reaperrr committed:
Allow Armaments to ignore custom AttackablePositions

reaperrr commented:
Updated.

No idea how the commit order got messed up (commit 3 should be commit 5), but that shouldn't have any effect as long as you don't test the first 5 commits individually.

reaperrr commented:
Updated.

`TargetPositionOffset` is now actually relative to body orientation. 
Used to fix the TS weed harvester's target position to match the voxel.

GraionDilach commented:
>OpenRA.Mods.Common/Traits/Health.cs:L108: [SP2000] Invalid spacing at the end of the line.

reaperrr commented:
Fixed style issue.

abcdefg30 commented:
Needs a rebase.

reaperrr commented:
Rebased.

pchote commented:
Can we add a height / offset for infantry too? You should be able to set it once on the default with something reasonable.

pchote commented:
Having actors not shoot at eachother's feet / wheels is a subtle but very nice polish improvement.

reaperrr commented:
@pchote: Fixed the factory offsets.

About infantry, I only gave them a 'ceiling' but left the target position at ground for now.
Two reasons:
1) TS' machine gun impact and anti-tank warhead explosions are visually designed for hitting flat surfaces, so while it looks good enough for vehicles and buildings, it looks rather weird on infantry.
2) Infantry prone sequences show them flat on the ground and - just like tick tanks - we'll need upgradable HitShapes + a tie to the prone logic to cover this properly, and I don't want to open that can of worms in this PR.

Mailaender commented:
I got:

```
Exception of type `System.InvalidOperationException`: Sequence contains no elements
  at System.Linq.Enumerable.First[TSource] (IEnumerable`1 source) [0x00067] in /home/abuild/rpmbuild/BUILD/mono-4.2.2/external/referencesource/System.Core/System/Linq/Enumerable.cs:930 
  at OpenRA.Mods.Common.Warheads.CreateEffectWarhead.GetDirectHit (OpenRA.World world, CPos cell, WPos pos, OpenRA.Actor firedBy, Boolean checkTargetType) <0x422b2680 + 0x0010f> in <filename unknown>:0 
  at OpenRA.Mods.Common.Warheads.CreateEffectWarhead.GetImpactType (OpenRA.World world, CPos cell, WPos pos, OpenRA.Actor firedBy) <0x422b24b0 + 0x000d7> in <filename unknown>:0 
  at OpenRA.Mods.Common.Warheads.CreateEffectWarhead.IsValidImpact (WPos pos, OpenRA.Actor firedBy) <0x422b23b0 + 0x00097> in <filename unknown>:0 
  at OpenRA.Mods.Common.Warheads.CreateEffectWarhead.DoImpact (Target target, OpenRA.Actor firedBy, IEnumerable`1 damageModifiers) <0x422b2050 + 0x000cf> in <filename unknown>:0 
  at OpenRA.GameRules.WeaponInfo+<Impact>c__AnonStorey1.<>m__0 () [0x00027] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/GameRules/WeaponInfo.cs:161 
  at OpenRA.GameRules.WeaponInfo.Impact (Target target, OpenRA.Actor firedBy, IEnumerable`1 damageModifiers) [0x00095] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/GameRules/WeaponInfo.cs:165 
  at OpenRA.Mods.Common.Activities.FallToEarth.Tick (OpenRA.Actor self) <0x42311290 + 0x0013b> in <filename unknown>:0 
  at OpenRA.Traits.ActivityUtils.RunActivity (OpenRA.Actor self, OpenRA.Activities.Activity act) [0x00024] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Traits/ActivityUtils.cs:37 
  at OpenRA.Actor.Tick () [0x00010] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Actor.cs:164 
  at OpenRA.World.Tick () [0x000c3] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/World.cs:331 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x00226] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:556 
  at OpenRA.Game.LogicTick () [0x00055] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:581 
  at OpenRA.Game.Loop () [0x000d4] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:708 
  at OpenRA.Game.Run () [0x00042] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:748 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:118 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:40
```

when shooting down an Orca Bomber with a SAM site. Not sure if it is related though.

reaperrr commented:
`CreateEffectWarhead` doesn't use Linq on bleed, so it must be caused by cb3a760.

I'm not sure under which circumstances the Sequence can contain no elements there, I couldn't reproduce the crash so far.

reaperrr commented:
Rebased and updated. I suspect that `var healthInfo = unit.Info.TraitInfos<HealthInfo>();` never returned `null`, so the following null check never triggered.

Mailaender commented:
Seems to work fine now. :white_check_mark:

reaperrr commented:
As requested by @pchote I have split this up into less intimidating, more reviewable chunks.

#11168 first, then this, then #11169.

reaperrr commented:
Updated.

GraionDilach commented:
Do you intend to use/have a usecase for the Footprint targetable position? I can understand why did you made the restriction, however, it seems really useless with it and all the cases where I'd even think about using it could be resolved with a bit of math/trial-and-error on CenterPosition offset even.

With all in mind, I probably would be more okay with dropping/removing it atm instead of merging it in it's unfinished state.

reaperrr commented:
> Do you intend to use/have a usecase for the Footprint targetable position?

Not really, I just preserved the old `TargetablePositions` from `Building` that way. Personally I agree that we probably don't need it, but maybe we're overlooking something. Could you chime in on this @pchote ?

GraionDilach commented:
Oh wait, now I remember what that was supposed to do which might even break on this PR: if a unit has lower range than the building width, the unit might be unable to reach the targeting position and thereby attack the building (say, melee units).

reaperrr commented:
You're right, and unfortunately it really breaks on this PR.
Setting the type to Footprint fixes the targeting, but then weapons aim at the first (top-left) cell of the footprint. This will need to be fixed, obviously.

reaperrr committed:
Introduce PositionClosestTo util

reaperrr committed:
Allow turrets and frontal-attacking actors to ignore AttackablePositions

..and face the actor's CenterPosition instead.

reaperrr commented:
Updated.

The underlying issue of what @GraionDilach pointed out is that on bleed, targeting and actual attacking aim at different position types. Using `TargetablePositions` for everything would have broken that separation, and using something hacky like first pos = targeting, 2nd pos attacking felt wrong, so in oder to keep the regression risk low and the separation intact, I decided to introduce an additional `IAttackablePositions` interface that basically works like `TargetablePositions`, but allows to leave the latter as-is and not cause any trouble there.

In addition to that, things like `Armament` no longer just aim at the first `AttackablePosition`, but rather the closest one to the source by default (unless this behavior is disabled). Later, when we have a Footprint HitShape, we can use this to reproduce D2k-like shooting at the closest cell of a building rather than the center.

obrakmann commented on OpenRA.Game/Traits/Target.cs:
```diff
@@ -165,6 +165,35 @@ public IEnumerable<WPos> Positions
+						if (attackablePositions.Any())
+						{
+							// TODO: Add the AttackablePositions of all traits implementing IAttackablePositions
+							return attackablePositions.First().AttackablePositions(actor);
```
This compiles, and seems to work in RA, at least.
```csharp
if (attackablePositions.Any())
{
	// TODO: Add the AttackablePositions of all traits implementing IAttackablePositions
	var target = this;
	return attackablePositions.SelectMany(ap => ap.AttackablePositions(target.actor));
}
```

That workaround is required because `Target` is a struct, not a class.

obrakmann commented on OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs:
```diff
@@ -58,6 +58,28 @@ public void RenderAfterWorld(WorldRenderer wr, Actor self)
 				healthInfo.Shape.DrawCombatOverlay(wr, wcr, self);
 
+			var attackablePositions = self.TraitsImplementing<IAttackablePositions>();
+			if (attackablePositions.Any())
```
You don't need this, `foreach` will handle the case just fine.

obrakmann commented on OpenRA.Mods.Common/HitShapes/Capsule.cs:
```diff
@@ -85,6 +85,13 @@ public WDist DistanceFromEdge(WVec v)
+				var positions = attackablePositions.First().AttackablePositions(actor);
+				if (positions.Any())
+					actorPos = positions.First();
+			}
```
Use `SelectMany` here as well, then use `PositionoClosestTo(actorPos)`? Using `First()` - twice! - here feels a bit dodgy.

Same thing in the hunk below.

reaperrr replied:
> Using First() - twice! - here feels a bit dodgy

Yeah, I wrote this before I came up with `PositionClosestTo`.

reaperrr commented:
Updated.

obrakmann commented:
ok, looks fine so far. Three more things to note, though:

1. Could you mention a use case for `Armament.IgnoreAttackablePositionOffset` in the commit message of the second-to-last commit, please?
2. Could you please describe (in its commit message) which problem the last commit is supposed to solve? Also, does that depend on the equivalent setting on `Armament` to be set as well? I imagine that otherwise the barrel of a turret could look one way while it would shoot the other way.
3. I'm not too sure if I like this `AttackablePositions` thing living on `Health`. Would `Targetable` work instead, for example?

GraionDilach commented:
Targetable doesn't care about the physical properties of the unit (hitshapes and stuff). The best solution could prolly be to decouple hitshapes and all related from Health into a new trait (especially if multiple health traits arrive which I guess @reaperrr do wants).

reaperrr commented:
> I'm not too sure if I like this AttackablePositions thing living on Health. Would Targetable work instead, for example?

@obrakmann I see your point, but it's like @GraionDilach said. This PR will be followed by a PR that allows multiple Health / HitZone traits per actor (splitting off things like `HitShapes` to a separate trait sounds like a good idea), and I don't want to mess with `Targetable`'s pure "boolean" nature.

I definitely want to complete the entire feature before the next release, so letting `AttackablePositions` live on `Health` should be acceptable as temporary solution.

pchote commented:
I agree with @obrakmann, putting this on health doesnt really sit right with me, even in the short term.  If you can wait a few days for me to recover from my work mission then i can look at this and possibly give some suggestions.

obrakmann commented:
Alright, so do I understand correctly that in the end it would live on a new trait that would combine both the hit shapes as well as the attackable positions? If so, :+1: 

(I would really prefer that "new" trait to be `Targetable`, it just makes sense context-wise, unless I'm missing something.)

GraionDilach commented:
I am really not a fan of expanding `Targetable` with this, because that would overcomplicate multiple active `Targetable` traits on actors without any actual benefit. The same reason I'd go with having HitShapes and positions on a third trait (multiple Health traits prolly also would complicate that usecase).

pchote commented:
We already have an interface and code in place for target positions, which very deliberately avoid touching Health.  Health is concerned with hitshapes because those determine how damage (from e.g. Spread) is applied.  This is only weakly related to the positions that other actors should target.

pchote commented:
Having separate positions for targeting vs attacking is rather crap - conceptually they are supposed to be the same thing. Can we fix that, instead?

reaperrr commented:
> Having separate positions for targeting vs attacking is rather crap - conceptually they are supposed to be the same thing. Can we fix that, instead?

If you have a different solution for the issue @GraionDilach pointed out a few days ago (melee units not being able to attack large buildings if their footprint doesn't allow the melee units to move close enough to the targetable position), then I don't mind.

pchote commented:
Buildings are supposed to expose a targetable position for each cell in their footprint, which is exposed via `Target.Positions`.  The targeting and attack code are both supposed to pick one of these points to attack.  If melee units are picking positions that are outside their range then that's a bug that needs to be fixed.

My different solution is to work out what broke in the logic above and then fix that :)

reaperrr commented:
The sheer amount of comments and changes this has gone through made it harder to review and a few dependencies still need to be implemented first, so I'll probably open a new, 'fresh' PR when I come back to this.
